File: Ruby/src/collatz.rb
def collatz_iteration(number)
  steps = 0
  while number != 1 do
    print("#{number} ")
    if number % 2 == 0
      number = number / 2
    else
      number = 3 * number + 1
    end
    steps = steps + 1
  end
  puts(number)
  steps
end

def collatz_recursive(number)
  print("#{number} ")
  if number == 1
    return
  end
  if number % 2 == 0
    collatz_recursive(number / 2)
  else
    collatz_recursive(3 * number + 1)
  end
end

input_number = 27
puts(collatz_iteration(input_number))
collatz_recursive(input_number)

File: R/src/collatz.R
collatzIteration <- function(number) {
  steps <- 0
  output <- ""
  while (number != 1) {
    output <- paste(output, number, sep = " ")
    if ((number %% 2) == 0) {
      number <- number / 2
    } else {
      number <- 3 * number + 1
    }
    steps <- steps + 1
  }
  output <- paste(output, number, sep = " ")
  print(output)
  return(steps)
}

collatzRecursive <- function(number) {
  cat(sprintf("%d ", number))
  if (number == 1) {
    return(0)
  }
  if (number %% 2 == 0) {
    collatzRecursive(number / 2)
  } else {
    collatzRecursive(3 * number + 1)
  }
}

inputNumber <- 27
print(collatzIteration(inputNumber))
collatzRecursive(inputNumber)
